File: biddo_client/lib/core/models/auction.dart
import 'package:biddo/core/controllers/image_picker.dart';
import 'package:get/get.dart';
import 'package:google_maps_flutter/google_maps_flutter.dart';
import 'package:image_picker/image_picker.dart';
import 'package:uuid/uuid.dart';
import 'package:google_maps_flutter/google_maps_flutter.dart' as google_map;
// ignore: library_prefixes
import './review.dart' as ReviewModel;

import 'account.dart';
import 'asset.dart';
import 'bid.dart';

var uuid = const Uuid();

enum AuctionProductCondition { newProduct, used }

class Auction {
  String id;
  String title;
  String description;
  AuctionProductCondition condition;
  Account? auctioneer;

  LatLng? location;
  String? locationPretty;
  List<Asset>? assets;
  List<Rx<Bid>> bids;
  List<ReviewModel.Review>? reviews;
  int? views;
  String? youtubeLink;

  String? acceptedBidId;
  DateTime? acceptedBidAt;
  int bidsCount;

  bool? hasCustomStartingPrice;
  double startingPrice;
  double? lastPrice;
  String? lastPriceCurrencyId;

  int? likesCount;

  String mainCategoryId;
  String subCategoryId;

  String? initialCurrencyId;

  DateTime? startAt;
  DateTime? startedAt;

  DateTime createdAt;
  DateTime? updatedAt;
  DateTime? expiresAt;
  DateTime? promotedAt;
  bool isActive;

  List<AuctionHistoryEvent>? historyEvents;

  Auction({
    String? id,
    required this.auctioneer,
    this.description = '',
    this.title = '',
    this.condition = AuctionProductCondition.newProduct,
    this.location,
    this.locationPretty,
    this.assets,
    this.bids = const [],
    this.reviews,
    this.likesCount = 0,
    this.initialCurrencyId,
    this.views,
    this.youtubeLink,
    this.acceptedBidId,
    this.acceptedBidAt,
    this.bidsCount = 0,
    this.hasCustomStartingPrice,
    this.startingPrice = 0,
    this.lastPrice,
    this.lastPriceCurrencyId,
    required this.mainCategoryId,
    required this.subCategoryId,
    required this.createdAt,
    this.expiresAt,
    this.updatedAt,
    this.promotedAt,
    this.startedAt,
    this.startAt,
    this.historyEvents,
    this.isActive = true,
  }) : id = id ?? uuid.v4();

  static Auction fromJSON(dynamic data) {
    var auctioneer =
        data['account'] != null ? Account.fromJSON(data['account']) : null;
    var locationLat = data['locationLat'];
    var locationLong = data['locationLong'];
    var locationLatLng = locationLat != null && locationLong != null
        ? LatLng(double.parse(locationLat.toString()),
            double.parse(locationLong.toString()))
        : null;

    // ignore: prefer_null_aware_operators
    var reviews = data['reviews'] != null
        ? data['reviews']
            .map<ReviewModel.Review>(
                (review) => ReviewModel.Review.fromJSON(review))
            .toList()
        : null;

    // It is active only if the auction expiresAt is greater than the current time
    var isActive = data['expiresAt'] != null
        ? DateTime.parse(data['expiresAt']).isAfter(DateTime.now())
        : true;

    return Auction(
      id: data['id'],
      title: data['title'],
      auctioneer: auctioneer,
      reviews: reviews,
      isActive: isActive,
      initialCurrencyId: data['initialCurrencyId'],
      condition: data['isNewItem']?.toString() == 'true'
          ? AuctionProductCondition.newProduct
          : AuctionProductCondition.used,
      description: data['description'] ?? '',
      locationPretty: data['locationPretty'] ?? '',
      location: locationLatLng,
      youtubeLink: data['youtubeLink'],
      startingPrice: data['startingPrice'] != null
          ? double.parse(data['startingPrice'].toString())
          : 0,
      lastPriceCurrencyId: data['lastPriceCurrencyId'],
      lastPrice: data['lastPrice'] != null
          ? double.parse(data['lastPrice'].toString())
          : 0,
      likesCount: data['likesCount'] != null
          ? int.parse(data['likesCount'].toString())
          : 0,
      mainCategoryId: data['mainCategoryId'],
      subCategoryId: data['subCategoryId'],
      acceptedBidId: data['acceptedBidId'],
      bidsCount: data['bidsCount'] ?? 0,
      views: data['views'] != null ? int.parse(data['views'].toString()) : null,
      hasCustomStartingPrice: data['hasCustomStartingPrice'] ?? false,
      assets: data['auctionAssets'] != null
          ? data['auctionAssets']
              .map<Asset>((asset) => Asset.fromJSON(asset))
              .toList()
          : [],
      bids: data['bids'] != null
          ? data['bids'].map<Rx<Bid>>((bid) => Bid.fromJSON(bid).obs).toList()
          : [],
      acceptedBidAt: data['acceptedBidAt'] != null
          ? DateTime.parse(data['acceptedBidAt'])
          : null,
      startAt: data['startAt'] != null ? DateTime.parse(data['startAt']) : null,
      startedAt:
          data['startedAt'] != null ? DateTime.parse(data['startedAt']) : null,
      createdAt: DateTime.parse(data['createdAt']),
      updatedAt: DateTime.parse(data['updatedAt']),
      promotedAt: data['promotedAt'] != null
          ? DateTime.parse(data['promotedAt'])
          : null,
      expiresAt:
          data['expiresAt'] != null ? DateTime.parse(data['expiresAt']) : null,
      historyEvents: data['auctionHistoryEvents'] != null
          ? data['auctionHistoryEvents']
              .map<AuctionHistoryEvent>(
                  (event) => AuctionHistoryEvent.fromJSON(event))
              .toList()
          : [],
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'title': title,
      'description': description,
      'auctionAssets': assets?.map((asset) => asset.toJson()).toList(),
      'acceptedBidId': acceptedBidId,
      'acceptedBidAt': acceptedBidAt,
      'bidsCount': bidsCount,
      'hasCustomStartingPrice': hasCustomStartingPrice,
      'startingPrice': startingPrice,
      'lastPrice': lastPrice,
      'mainCategoryId': mainCategoryId,
      'subCategoryId': subCategoryId,
      'isActive': isActive,
      'createdAt': createdAt.toString(),
      'updatedAt': updatedAt?.toString(),
    };
  }
}

class AuctionHistoryEvent {
  String id;
  String type;
  dynamic details;

  DateTime createdAt;

  AuctionHistoryEvent({
    required this.id,
    required this.type,
    required this.details,
    required this.createdAt,
  });

  static AuctionHistoryEvent fromJSON(dynamic data) {
    return AuctionHistoryEvent(
      id: data['id'],
      type: data['type'],
      details: data['details'],
      createdAt: DateTime.parse(data['createdAt']),
    );
  }
}

class CreateUpdateAuctionParams {
  String? id;
  String title;
  String location;
  String? description;
  bool hasCustomStartingPrice;
  String mainCategoryId;
  String subCategoryId;
  double price;
  String condition;
  String? youtubeLink;
  String? initialCurrencyId;

  List<String>? assetsToKeep;
  DateTime? startAt;
  DateTime? expiresAt;

  google_map.LatLng latLng;
  List<AssetEntityPath>? galleryAssets;
  List<XFile>? cameraAssets;

  CreateUpdateAuctionParams({
    this.id,
    required this.title,
    required this.location,
    required this.price,
    required this.hasCustomStartingPrice,
    required this.latLng,
    required this.mainCategoryId,
    required this.subCategoryId,
    this.description,
    this.startAt,
    this.youtubeLink,
    this.initialCurrencyId,
    this.expiresAt,
    this.galleryAssets,
    this.cameraAssets,
    this.assetsToKeep,
    this.condition = 'new',
  });
}

File: core/domain/auction.ts
import { makeAutoObservable } from 'mobx'
import { Account } from './account'
import { Asset } from './asset'
import { Bid } from './bid'
import { Review } from './review'

export enum AuctionProductCondition {
  newProduct,
  used,
}

export class Auction {
  id: string
  title: string
  description: string
  condition: AuctionProductCondition
  auctioneer?: Account
  youtubeLink?: string

  location: LatLng
  locationPretty: string
  assets?: Asset[]
  bids: Bid[]
  reviews: Review[]
  views?: number

  mainCategoryId: string
  subCategoryId: string

  acceptedBidAt?: Date
  acceptedBidId?: string
  bidsCount?: number

  hasCustomStartingPrice?: boolean
  startingPrice?: number
  lastPrice?: number
  likesCount?: number

  lastPriceCurrencyId?: string
  initialCurrencyId?: string

  whoPaysForShipment?: string

  startAt?: Date
  startedAt?: Date

  expiresAt?: Date
  createdAt: Date
  updatedAt: Date
  promotedAt?: Date
  isActive: boolean

  historyEvents?: AuctionHistoryEvent[]

  constructor(params: Record<string, unknown> = {}) {
    this.id = params.id as string
    this.title = params.title as string
    this.description = params.description as string
    this.location = params.location as LatLng
    this.locationPretty = params.locationPretty as string
    this.auctioneer = params.auctioneer as Account
    this.assets = params.assets as Asset[]
    this.bids = params.bids as Bid[]
    this.condition = params.condition as AuctionProductCondition
    this.reviews = params.reviews as Review[]
    this.views = params.views as number
    this.isActive = (params.isActive as boolean) ?? true
    this.acceptedBidAt = params.acceptedBidAt as Date
    this.acceptedBidId = params.acceptedBidId as string
    this.bidsCount = params.bidsCount as number
    this.hasCustomStartingPrice = params.hasCustomStartingPrice as boolean
    this.startingPrice = params.startingPrice as number
    this.youtubeLink = params.youtubeLink as string
    this.lastPrice = params.lastPrice as number
    this.mainCategoryId = params.mainCategoryId as string
    this.subCategoryId = params.subCategoryId as string
    this.whoPaysForShipment = params.whoPaysForShipment as string
    this.expiresAt = params.expiresAt as Date
    this.createdAt = params.createdAt as Date
    this.updatedAt = params.updatedAt as Date
    this.promotedAt = params.promotedAt as Date
    this.startAt = params.startAt as Date
    this.startedAt = params.startedAt as Date
    this.historyEvents = params.historyEvents as AuctionHistoryEvent[]
    this.lastPriceCurrencyId = params.lastPriceCurrencyId as string
    this.initialCurrencyId = params.initialCurrencyId as string

    makeAutoObservable(this)
  }

  static fromJSON(data: Record<string, unknown> = {}) {
    const locationLat = data.locationLat as number
    const locationLong = data.locationLong as number

    const locationLatLng =
      locationLat != null && locationLong != null
        ? new LatLng({ lat: locationLat, lng: locationLong })
        : null
    const auctioneer = data.account
      ? Account.fromJSON(data.account as Record<string, unknown>)
      : null

    const reviews = data.reviews ? (data.reviews as Partial<Review>[]).map(Review.fromJSON) : []
    const bids = data.bids ? (data.bids as Partial<Bid>[]).map(Bid.fromJSON) : []

    let assets = data.auctionAssets
      ? (data.auctionAssets as Partial<Asset>[]).map(Asset.fromJSON)
      : []

    if (!assets?.length && Array.isArray(data.assets) && data.assets?.length) {
      assets = (data.assets as Partial<Asset>[]).map(Asset.fromJSON)
    }

    let locationPretty = data.locationPretty ?? ''
    if (!locationPretty || locationPretty === 'undefined') {
      locationPretty = ''
    }

    const historyEvents = data.auctionHistoryEvents
      ? (data.auctionHistoryEvents as Partial<AuctionHistoryEvent>[]).map(
          AuctionHistoryEvent.fromJSON
        )
      : []

    return new Auction({
      id: data.id,
      title: data.title,
      condition: data.isNewItem ? AuctionProductCondition.newProduct : AuctionProductCondition.used,
      description: data.description,
      youtubeLink: data.youtubeLink,
      location: locationLatLng,
      lastPriceCurrencyId: data.lastPriceCurrencyId,
      initialCurrencyId: data.initialCurrencyId,
      mainCategoryId: data.mainCategoryId,
      subCategoryId: data.subCategoryId,
      locationPretty,
      auctioneer,
      assets,
      bids,
      reviews,
      historyEvents,
      views: data.views,
      countdownPreference: data.countdownPreference,
      countdownStarted: data.countdownStarted,
      isActive: data.isActive,
      acceptedBidAt: data.acceptedBidAt ? new Date(data.acceptedBidAt.toString()) : null,
      acceptedBidId: data.acceptedBidId,
      bidsCount: data.bidsCount,
      reactivatedAt: data.reactivatedAt ? new Date(data.reactivatedAt.toString()) : null,
      reactivationCount: data.reactivationCount ?? 0,
      hasCustomStartingPrice: data.hasCustomStartingPrice,
      startingPrice: data.startingPrice ? parseFloat(data.startingPrice.toString()) : 0,
      lastPrice: data.lastPrice ? parseFloat(data.lastPrice.toString()) : 0,
      whoPaysForShipment: data.whoPaysForShipment,
      expiresAt: data.expiresAt ? new Date(data.expiresAt.toString()) : null,
      startAt: data.startAt ? new Date(data.startAt.toString()) : null,
      startedAt: data.startedAt ? new Date(data.startedAt.toString()) : null,

      createdAt: data.createdAt ? new Date(data.createdAt.toString()) : null,
      updatedAt: data.updatedAt ? new Date(data.updatedAt.toString()) : null,
      promotedAt: data.promotedAt ? new Date(data.promotedAt.toString()) : null,
    })
  }
}

export class LatLng {
  lat: number
  lng: number

  constructor(params: Record<string, number> = {}) {
    this.lat = params.lat
    this.lng = params.lng
  }
}

export class AuctionHistoryEvent {
  id: string
  type: string
  details: Record<string, string>
  createdAt: Date

  constructor(params: Record<string, unknown> = {}) {
    this.id = params.id as string
    this.type = params.type as string
    this.details = params.details as Record<string, string>
    this.createdAt = params.createdAt as Date
  }

  static fromJSON(data: Record<string, unknown> = {}) {
    return new AuctionHistoryEvent({
      id: data.id,
      type: data.type,
      details: data.details,
      createdAt: data.createdAt ? new Date(data.createdAt.toString()) : null,
    })
  }
}
